{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Building `dms-view` datasets for [Sourisseau *et al.*, 2019](https://research.fhcrc.org/content/dam/stripe/bloom/labfiles/publications/Sourisseau2019.pdf)\n",
    "\n",
    "This jupyter notebook builds a `dms-view` datafile for the Deep Mutational Scanning (DMS; `Sourisseau2019_DMS.csv`) and the Mutational Antigenic Profiling (MAP; `Sourisseau2019_MAP.csv`) of Zika Envelope protein.\n",
    "The data is scraped from the [paper repo](https://github.com/jbloomlab/ZIKV_DMS_with_EvansLab). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## notebook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import entropy\n",
    "from Bio import SeqIO\n",
    "from phydmslib.constants import *\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create new protein structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_chains = ['C', 'E']\n",
    "with open(f'5ire_{\"_\".join(target_chains)}.pdb', 'w') as outfile:\n",
    "    with open('5ire.pdb', 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('ATOM') and line.split()[4] not in target_chains:\n",
    "                pass\n",
    "            else:\n",
    "                outfile.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deep Mutational Scanning data\n",
    "\n",
    "The DMS data will show two \"conditions\": raw preferences, and rescaled preferences. \n",
    "For each \"condition\", the dot plot will show either the entropy, n effective, or the RSA and the logoplot will show the prefs value, the mutational effect (shared between conditions), and the natural frequencies (shared between conditions)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### paths to datafiles on github"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "muteffects_fname = ('https://raw.githubusercontent.com/jbloomlab/ZIKV_DMS_with_EvansLab/master/results/'\n",
    "                    'muteffects/unscaled_muteffects.csv')\n",
    "prefs_fname = ('https://raw.githubusercontent.com/jbloomlab/ZIKV_DMS_with_EvansLab/master/results/'\n",
    "               'prefs/summary_avgprefs.csv')\n",
    "rescaledprefs_fname = ('https://raw.githubusercontent.com/jbloomlab/ZIKV_DMS_with_EvansLab/master/results/'\n",
    "                       'prefs/rescaled_prefs.csv')\n",
    "sitesummary_fname= ('https://raw.githubusercontent.com/jbloomlab/ZIKV_DMS_with_EvansLab/master/results/'\n",
    "                    'struct_props/struct_props_mut_tol.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>site_RSA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.030457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>0.244526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>0.023952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>0.142132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>5</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site  site_RSA\n",
       "0      1  0.030457\n",
       "4      2  0.244526\n",
       "8      3  0.023952\n",
       "12     4  0.142132\n",
       "16     5  0.625000"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in RSA and entropy/n effective \n",
    "RSA = (pd.read_csv(sitesummary_fname).query(\"pdb == '5ire'\")\n",
    "       [['site', 'RSA', 'mutational_tolerance_measure', 'mutational_tolerance']]\n",
    "       .rename(columns={'RSA': 'site_RSA'})\n",
    "       .drop(columns=['mutational_tolerance_measure', 'mutational_tolerance']))\n",
    "RSA.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mutational effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_mutational effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>A</td>\n",
       "      <td>-5.046030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.308696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>-3.884530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>-4.290569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>F</td>\n",
       "      <td>-4.370049</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site wildtype mutation  mut_mutational effect\n",
       "0     1        I        A              -5.046030\n",
       "1     1        I        C              -4.308696\n",
       "2     1        I        D              -3.884530\n",
       "3     1        I        E              -4.290569\n",
       "4     1        I        F              -4.370049"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_effects = (pd.read_csv(muteffects_fname)\n",
    "               .drop(columns=['mutation', 'effect'])\n",
    "               .rename(columns={'mutant': 'mutation', \n",
    "                                'log2effect': 'mut_mutational effect'}))\n",
    "mut_effects.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### natural frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "translate ={}\n",
    "for codon in CODON_TO_INDEX.items():\n",
    "    translate[codon[0]] = INDEX_TO_AA[CODON_TO_AA[codon[1]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_natural frequencies</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site mutation  mut_natural frequencies\n",
       "0     1        A                      0.0\n",
       "1     2        A                      0.0\n",
       "2     3        A                      0.0\n",
       "3     4        A                      0.0\n",
       "4     5        A                      0.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in data\n",
    "with open('E_alignment.fasta', \"r\") as handle:\n",
    "    alignment = [(x.id, str(x.seq).strip()) for x in SeqIO.parse(handle, \"fasta\")]\n",
    "\n",
    "seqlength = len(alignment[0][1]) // 3\n",
    "nat = {k: [0 for x in range(seqlength)] for k in list(INDEX_TO_AA.values())}\n",
    "\n",
    "# count amino acid frequencies\n",
    "for seq in alignment:\n",
    "    for i in range(seqlength):\n",
    "        codon = seq[1][3 * i: 3 * i + 3]\n",
    "        if codon != '---':\n",
    "            nat[translate[codon]][i] += 1\n",
    "nat = pd.DataFrame(nat)\n",
    "\n",
    "# Normalize the dataframe\n",
    "assert not np.any(nat.sum(axis=1) == 0), (\"Attempting to normalize a \"\n",
    "                                            \"site by an amino acid count\"\n",
    "                                            \" of zero. Does the alignment\"\n",
    "                                            \" have an all gap column?\")\n",
    "nat = nat.div(nat.sum(axis=1), axis=0)\n",
    "assert np.allclose(nat.sum(axis=1), 1, atol=0.005)\n",
    "nat[\"site\"] = [x+1 for x in range(len(nat))]\n",
    "\n",
    "# convert to dms-view format\n",
    "nat = pd.melt(nat, id_vars='site', var_name='mutation', value_name='mut_natural frequencies')\n",
    "nat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### preferences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>condition</th>\n",
       "      <th>site_entropy</th>\n",
       "      <th>site_n effective</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_pref</th>\n",
       "      <th>site_RSA</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mut_mutational effect</th>\n",
       "      <th>mut_natural frequencies</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>raw preferences</td>\n",
       "      <td>1.809948</td>\n",
       "      <td>3.506295</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005438</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.046030</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>raw preferences</td>\n",
       "      <td>1.809948</td>\n",
       "      <td>3.506295</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005438</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.046030</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rescaled preferences</td>\n",
       "      <td>1.317192</td>\n",
       "      <td>2.491807</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-8.325950</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rescaled preferences</td>\n",
       "      <td>1.317192</td>\n",
       "      <td>2.491807</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-8.325950</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>raw preferences</td>\n",
       "      <td>1.809948</td>\n",
       "      <td>3.506295</td>\n",
       "      <td>C</td>\n",
       "      <td>0.009065</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-4.308696</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site             condition  site_entropy  site_n effective mutation  \\\n",
       "0     1       raw preferences      1.809948          3.506295        A   \n",
       "1     1       raw preferences      1.809948          3.506295        A   \n",
       "2     1  rescaled preferences      1.317192          2.491807        A   \n",
       "3     1  rescaled preferences      1.317192          2.491807        A   \n",
       "4     1       raw preferences      1.809948          3.506295        C   \n",
       "\n",
       "   mut_pref  site_RSA wildtype  mut_mutational effect  mut_natural frequencies  \n",
       "0  0.005438  0.030457        I              -5.046030                      0.0  \n",
       "1  0.005438  0.030457        I              -5.046030                      0.0  \n",
       "2  0.000505  0.030457        I              -8.325950                      0.0  \n",
       "3  0.000505  0.030457        I              -8.325950                      0.0  \n",
       "4  0.009065  0.030457        I              -4.308696                      0.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amino_acids = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L','M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "beta = 1.65\n",
    "df = (pd.concat([(pd.read_csv(fname)[['site'] + amino_acids])\n",
    "                    .assign(condition=condition)\n",
    "                    for condition, fname in [('raw preferences', prefs_fname),\n",
    "                                             ('rescaled preferences', rescaledprefs_fname)]\n",
    "                   ]))\n",
    "df['site_entropy'] = df[amino_acids].apply(lambda x: entropy(x), axis=1)\n",
    "df['site_n effective'] = df['site_entropy'].apply(lambda x: 2**x)\n",
    "df = pd.melt(df, \n",
    "                id_vars=['site', 'condition', 'site_entropy', 'site_n effective'], \n",
    "                var_name='mutation', \n",
    "                value_name='mut_pref')\n",
    "df = pd.merge(df, RSA, on=['site'])\n",
    "df = pd.merge(df, mut_effects, on=['site', 'mutation'])\n",
    "df = pd.merge(df, nat, on=['site', 'mutation'])\n",
    "\n",
    "# rescale mutational effects\n",
    "df['mut_mutational effect'] = (df[['mut_mutational effect', 'condition']]\n",
    "                               .apply(lambda x: \n",
    "                                      x[0] * beta if x[1] == 'rescaled preferences' else x[0], axis=1))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>condition</th>\n",
       "      <th>site_entropy</th>\n",
       "      <th>site_n effective</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_pref</th>\n",
       "      <th>site_RSA</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mut_mutational effect</th>\n",
       "      <th>mut_natural frequencies</th>\n",
       "      <th>protein_chain</th>\n",
       "      <th>protein_site</th>\n",
       "      <th>label_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>raw preferences</td>\n",
       "      <td>1.809948</td>\n",
       "      <td>3.506295</td>\n",
       "      <td>A</td>\n",
       "      <td>0.005438</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-5.046030</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C E</td>\n",
       "      <td>1</td>\n",
       "      <td>I 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rescaled preferences</td>\n",
       "      <td>1.317192</td>\n",
       "      <td>2.491807</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-8.325950</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C E</td>\n",
       "      <td>1</td>\n",
       "      <td>I 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>raw preferences</td>\n",
       "      <td>1.809948</td>\n",
       "      <td>3.506295</td>\n",
       "      <td>C</td>\n",
       "      <td>0.009065</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-4.308696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C E</td>\n",
       "      <td>1</td>\n",
       "      <td>I 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>rescaled preferences</td>\n",
       "      <td>1.317192</td>\n",
       "      <td>2.491807</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-7.109348</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C E</td>\n",
       "      <td>1</td>\n",
       "      <td>I 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>raw preferences</td>\n",
       "      <td>1.809948</td>\n",
       "      <td>3.506295</td>\n",
       "      <td>D</td>\n",
       "      <td>0.012163</td>\n",
       "      <td>0.030457</td>\n",
       "      <td>I</td>\n",
       "      <td>-3.884530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C E</td>\n",
       "      <td>1</td>\n",
       "      <td>I 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   site             condition  site_entropy  site_n effective mutation  \\\n",
       "0     1       raw preferences      1.809948          3.506295        A   \n",
       "2     1  rescaled preferences      1.317192          2.491807        A   \n",
       "4     1       raw preferences      1.809948          3.506295        C   \n",
       "6     1  rescaled preferences      1.317192          2.491807        C   \n",
       "8     1       raw preferences      1.809948          3.506295        D   \n",
       "\n",
       "   mut_pref  site_RSA wildtype  mut_mutational effect  \\\n",
       "0  0.005438  0.030457        I              -5.046030   \n",
       "2  0.000505  0.030457        I              -8.325950   \n",
       "4  0.009065  0.030457        I              -4.308696   \n",
       "6  0.001174  0.030457        I              -7.109348   \n",
       "8  0.012163  0.030457        I              -3.884530   \n",
       "\n",
       "   mut_natural frequencies protein_chain  protein_site label_site  \n",
       "0                      0.0           C E             1        I 1  \n",
       "2                      0.0           C E             1        I 1  \n",
       "4                      0.0           C E             1        I 1  \n",
       "6                      0.0           C E             1        I 1  \n",
       "8                      0.0           C E             1        I 1  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['protein_chain'] = f\"{' '.join(target_chains)}\"\n",
    "df['protein_site'] = df['site']\n",
    "df['label_site'] = df[['wildtype', 'site']].apply(lambda x: f'{x[0]} {x[1]}', axis=1)\n",
    "df = df.drop_duplicates()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('Sourisseau2019_DMS.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutational Antigenic Profiling Data\n",
    "\n",
    "The MAP data will show two \"conditions\": antibody ZKA185 and antibody ZKA64. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### differential selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = ['ZKA64', 'ZKA185']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mut_positive diffsel</th>\n",
       "      <th>site_positive diffsel</th>\n",
       "      <th>site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_diffsel</th>\n",
       "      <th>site_abs diffsel</th>\n",
       "      <th>site_negative diffsel</th>\n",
       "      <th>site_max diffsel</th>\n",
       "      <th>site_min diffsel</th>\n",
       "      <th>condition</th>\n",
       "      <th>label_site</th>\n",
       "      <th>protein_chain</th>\n",
       "      <th>protein_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.664267</td>\n",
       "      <td>40.671283</td>\n",
       "      <td>333</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>6.442699</td>\n",
       "      <td>41.707742</td>\n",
       "      <td>-1.036459</td>\n",
       "      <td>6.442699</td>\n",
       "      <td>-0.542144</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>A 333</td>\n",
       "      <td>A</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7.384974</td>\n",
       "      <td>40.671283</td>\n",
       "      <td>333</td>\n",
       "      <td>A</td>\n",
       "      <td>E</td>\n",
       "      <td>5.739223</td>\n",
       "      <td>41.707742</td>\n",
       "      <td>-1.036459</td>\n",
       "      <td>6.442699</td>\n",
       "      <td>-0.542144</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>A 333</td>\n",
       "      <td>A</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7.262515</td>\n",
       "      <td>40.671283</td>\n",
       "      <td>333</td>\n",
       "      <td>A</td>\n",
       "      <td>R</td>\n",
       "      <td>4.704770</td>\n",
       "      <td>41.707742</td>\n",
       "      <td>-1.036459</td>\n",
       "      <td>6.442699</td>\n",
       "      <td>-0.542144</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>A 333</td>\n",
       "      <td>A</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7.177435</td>\n",
       "      <td>40.671283</td>\n",
       "      <td>333</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>4.428613</td>\n",
       "      <td>41.707742</td>\n",
       "      <td>-1.036459</td>\n",
       "      <td>6.442699</td>\n",
       "      <td>-0.542144</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>A 333</td>\n",
       "      <td>A</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.764480</td>\n",
       "      <td>40.671283</td>\n",
       "      <td>333</td>\n",
       "      <td>A</td>\n",
       "      <td>L</td>\n",
       "      <td>4.418945</td>\n",
       "      <td>41.707742</td>\n",
       "      <td>-1.036459</td>\n",
       "      <td>6.442699</td>\n",
       "      <td>-0.542144</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>A 333</td>\n",
       "      <td>A</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mut_positive diffsel  site_positive diffsel  site wildtype mutation  \\\n",
       "0              7.664267              40.671283   333        A        T   \n",
       "1              7.384974              40.671283   333        A        E   \n",
       "2              7.262515              40.671283   333        A        R   \n",
       "3              7.177435              40.671283   333        A        V   \n",
       "4              5.764480              40.671283   333        A        L   \n",
       "\n",
       "   mut_diffsel  site_abs diffsel  site_negative diffsel  site_max diffsel  \\\n",
       "0     6.442699         41.707742              -1.036459          6.442699   \n",
       "1     5.739223         41.707742              -1.036459          6.442699   \n",
       "2     4.704770         41.707742              -1.036459          6.442699   \n",
       "3     4.428613         41.707742              -1.036459          6.442699   \n",
       "4     4.418945         41.707742              -1.036459          6.442699   \n",
       "\n",
       "   site_min diffsel condition label_site protein_chain  protein_site  \n",
       "0         -0.542144     ZKA64      A 333             A           333  \n",
       "1         -0.542144     ZKA64      A 333             A           333  \n",
       "2         -0.542144     ZKA64      A 333             A           333  \n",
       "3         -0.542144     ZKA64      A 333             A           333  \n",
       "4         -0.542144     ZKA64      A 333             A           333  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = []\n",
    "\n",
    "# process data from the two antibodies\n",
    "for condition in conditions:\n",
    "    # mutation differential selection\n",
    "    mut = (pd.read_csv(('https://raw.githubusercontent.com/jbloomlab/ZIKV_DMS_with_EvansLab/master/results/'\n",
    "                       f'diffsel/summary_{condition}-meanmutdiffsel.csv'))\n",
    "           .rename(columns={'mutdiffsel': 'mut_diffsel'}))\n",
    "\n",
    "    # site differential selection\n",
    "    site = pd.read_csv(('https://raw.githubusercontent.com/jbloomlab/ZIKV_DMS_with_EvansLab/master/results/'\n",
    "                       f'diffsel/summary_{condition}-meansitediffsel.csv'))\n",
    "    site.columns = [f\"site_{' '.join(x.split('_'))}\" if x!='site' else x for x in site.columns.values]\n",
    "\n",
    "    # all together\n",
    "    tog = pd.merge(mut, site, on='site')\n",
    "    tog['condition'] = condition\n",
    "    df.append(tog)\n",
    "df = pd.concat(df)\n",
    "\n",
    "# add in protein information and site information\n",
    "df['label_site'] = df[['wildtype', 'site']].apply(lambda x: f'{x[0]} {x[1]}', axis=1)\n",
    "df['protein_chain'] = 'A'\n",
    "df['protein_site'] = df['site']\n",
    "\n",
    "# calculate positive differential selection\n",
    "df['mut_positive diffsel'] = mut['mut_diffsel'].clip(lower=0)\n",
    "\n",
    "# reorder the columns\n",
    "firstmut = 'mut_positive diffsel'\n",
    "firstsite = 'site_positive diffsel'\n",
    "cols = list(df.columns.values) \n",
    "cols.remove(firstmut) \n",
    "cols.remove(firstsite) \n",
    "df = df[[firstmut, firstsite] + cols]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('Sourisseau2019_MAP.csv', index=Fal)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
